clients/sov-celestia/src/client_state/mod.rs

```diff
@@ -23,16 +27,47 @@ pub use update_client::*;
 /// implement traits from `ibc::core::client::context` on the `ClientState`
 /// type.
 #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
-#[derive(Clone, Debug, PartialEq, derive_more::From)]
-pub struct ClientState(SovTmClientState);
+#[derive(Clone, Debug, PartialEq)]
+pub enum ClientState {
+    Native {
+        state: SovTmClientState,
+    },
+    Wasm {
```

Comment:
This approach might be too general, assuming if this `ClientState` type is used at places other than verifying counterparty client state. For example, if the client state is also used for local query, it should never return a Wasm-wrapped client state.

Also, we should not allow the two types to be interchangeable when verifying proofs. If it is first verified as a Wasm-wrapped client during connection open init, then it should always remain a Wasm-wrapped client, and vice versa.

Reply:
I agree that this allows creating clients with WASM wrapped types. But I don't see any other way to fix this - other than making ConnectionHandshake take an extra initialization parameter - that determines if the counterparty is actually a wasm-08 proxy client.